### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by gemmellr**
Provides AMQP 1.0 producers and consumers via a bridging layer implementing the event bus producer and consumer APIs over the top of the vertx-proton module.

Note that the docs and examples are still outstanding, I will work on those next week (Tuesday onwards), but as I discussed with Julien I wanted to get the PR raised now for the current code itself.

Robbie

### CODE REVIEW

**Comment by gemmellr** [comment]
Made a start on docs in-tree, and sender/receiver examples at: https://github.com/vert-x3/vertx-examples/tree/amqp-bridge-examples/amqp-bridge-examples

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    <license>
+      <name>The Apache Software License, Version 2.0</name>
+      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
+      <distribution>repo</distribution>
+    </license>
+  </licenses>
+
+  <properties>
+    <stack.version>3.3.0-SNAPSHOT</stack.version>
+    <vertx-proton.version>3.3.0-SNAPSHOT</vertx-proton.version>
```
Comment (vietj) [inline pom.xml]: this snapshot version should be handled in vertx-dependencies artifact

File: src/main/java/io/vertx/amqp/bridge/Bridge.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.eventbus.MessageConsumer;
+import io.vertx.core.eventbus.MessageProducer;
+
+/**
+ * Vert.x AMQP Bridge. Facilitates sending and receiving AMQP messages.
+ */
+@VertxGen
+public interface Bridge {
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/Bridge.java]: should be renamed `AmqpBridge` or `AMQPBridge`

File: src/main/java/io/vertx/amqp/bridge/BridgeOptions.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+import io.vertx.core.net.ProxyOptions;
+import io.vertx.core.net.SSLEngineOptions;
+import io.vertx.core.net.TrustOptions;
+import io.vertx.proton.ProtonClientOptions;
+
+/**
+ * Options for configuring {@link io.vertx.amqp.bridge.Bridge}.
+ */
+@DataObject(generateConverter = true, inheritConverter = true)
+public class BridgeOptions extends ProtonClientOptions {
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/BridgeOptions.java]: should be renamed to `AmqpBridgeOptions` or `AMQPBridgeOptions`

File: src/main/java/io/vertx/amqp/bridge/MessageHelper.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+* limitations under the License.
+*/
+package io.vertx.amqp.bridge;
+
+/**
+ * Helper for dealing with the various sections, and their constituent elements, that form the
+ * <a href="http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format">AMQP
+ * message</a>.
+ */
+public class MessageHelper {
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/MessageHelper.java]: imho should be renamed to `AmqpConstants` or `MessageConstants`

File: src/main/java/io/vertx/amqp/bridge/impl/AmqpConsumerImpl.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+import io.vertx.core.eventbus.MessageConsumer;
+import io.vertx.core.eventbus.impl.BodyReadStream;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.core.streams.ReadStream;
+import io.vertx.proton.ProtonConnection;
+import io.vertx.proton.ProtonReceiver;
+
+public class AmqpConsumerImpl implements MessageConsumer<JsonObject> {
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/impl/AmqpConsumerImpl.java:35]: AmqpConsumerImpl can be accessed from non Vert.x thread so its members that are accessed from the Vert.x thread (proton callbacks) should be visible from both threads.

So we need to have some synchronization in the consumer for this. Have a look in the original implementation of `MessageConsumer` (HandlerRegistration).

File: src/main/java/io/vertx/amqp/bridge/impl/AmqpProducerImpl.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import io.vertx.core.Handler;
+import io.vertx.core.VertxException;
+import io.vertx.core.eventbus.DeliveryOptions;
+import io.vertx.core.eventbus.Message;
+import io.vertx.core.eventbus.MessageProducer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.proton.ProtonConnection;
+import io.vertx.proton.ProtonSender;
+
+public class AmqpProducerImpl implements MessageProducer<JsonObject> {
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/impl/AmqpProducerImpl.java:29]: Same comment than AmqpConsumerImpl about synchronization. For example the drainHandler could be set from a non Vert.x thread (a main) but the callback for draining drain would likely be done by a Vert.x thread (proton).

File: src/main/java/io/vertx/amqp/bridge/impl/BridgeImpl.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+import io.vertx.proton.ProtonReceiver;
+
+public class BridgeImpl implements Bridge {
+
+  private ProtonClient client;
+  private ProtonConnection connection;
+  private ProtonReceiver replyToConsumer;
+  private String replyToConsumerAddress;
+  private AmqpProducerImpl replySender;
+  private Map<String, Handler<?>> replyToMapping = new HashMap<>();
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/impl/BridgeImpl.java]: should it be a ConcurrentHashMap ?

File: src/main/java/io/vertx/amqp/bridge/impl/BridgeMetaDataSupportImpl.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  public static final Symbol PRODUCT_KEY = Symbol.valueOf("product");
+  public static final String VERSION;
+  public static final Symbol VERSION_KEY = Symbol.valueOf("version");
+
+  static {
+    String version = "unknown";
+    try {
+      InputStream in = null;
+      String path = BridgeMetaDataSupportImpl.class.getPackage().getName().replace(".", "/");
+      if ((in = BridgeMetaDataSupportImpl.class.getResourceAsStream("/" + path + "/version.txt")) != null) {
```
Comment (vietj) [inline src/main/java/io/vertx/amqp/bridge/impl/BridgeMetaDataSupportImpl.java]: `in` should be closed

**Comment by gemmellr** [comment]
@vietj I have addressed the issues from your review and our subsequent discussions, can you take another look? Thanks!

**Comment by gemmellr** [comment]
I've also added a related PR for the examples repo:
https://github.com/vert-x3/vertx-examples/pull/141

**Pull request merged**